# Business Metrics

Anggota Team 4:
*   Akhmad Yuzfa Salvian Idris
*   Arief Rahman Hakim
*   Bernardus Valentino
*   Milenia Nadia Afifah Puspitasari
*   Sean Frederic Wijaya

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('Travel.csv')
df

CustomerID  ProdTaken   Age    TypeofContact  CityTier  DurationOfPitch  \
0         200000          1  41.0     Self Enquiry         3              6.0   
1         200001          0  49.0  Company Invited         1             14.0   
2         200002          1  37.0     Self Enquiry         1              8.0   
3         200003          0  33.0  Company Invited         1              9.0   
4         200004          0   NaN     Self Enquiry         1              8.0   
...          ...        ...   ...              ...       ...              ...   
4883      204883          1  49.0     Self Enquiry         3              9.0   
4884      204884          1  28.0  Company Invited         1             31.0   
4885      204885          1  52.0     Self Enquiry         3             17.0   
4886      204886          1  19.0     Self Enquiry         3             16.0   
4887      204887          1  36.0     Self Enquiry         1             14.0   

          Occupation  Gender  NumberOfPersonVisiting  NumberOfFollowups  \
0           Salaried  Female                       3                3.0   
1           Salaried    Male                       3                4.0   
2        Free Lancer    Male                       3                4.0   
3           Salaried  Female                       2                3.0   
4     Small Business    Male                       2                3.0   
...              ...     ...                     ...                ...   
4883  Small Business    Male                       3                5.0   
4884        Salaried    Male                       4                5.0   
4885        Salaried  Female                       4                4.0   
4886  Small Business    Male                       3                4.0   
4887        Salaried    Male                       4                4.0   

     ProductPitched  PreferredPropertyStar MaritalStatus  NumberOfTrips  \
0            Deluxe                    3.0        Single            1.0   
1            Deluxe                    4.0      Divorced            2.0   
2             Basic                    3.0        Single            7.0   
3             Basic                    3.0      Divorced            2.0   
4             Basic                    4.0      Divorced            1.0   
...             ...                    ...           ...            ...   
4883         Deluxe                    4.0     Unmarried            2.0   
4884          Basic                    3.0        Single            3.0   
4885       Standard                    4.0       Married            7.0   
4886          Basic                    3.0        Single            3.0   
4887          Basic                    4.0     Unmarried            3.0   

      Passport  PitchSatisfactionScore  OwnCar  NumberOfChildrenVisiting  \
0            1                       2       1                       0.0   
1            0                       3       1                       2.0   
2            1                       3       0                       0.0   
3            1                       5       1                       1.0   
4            0                       5       1                       0.0   
...        ...                     ...     ...                       ...   
4883         1                       1       1                       1.0   
4884         1                       3       1                       2.0   
4885         0                       1       1                       3.0   
4886         0                       5       0                       2.0   
4887         1                       3       1                       2.0   

         Designation  MonthlyIncome  
0            Manager        20993.0  
1            Manager        20130.0  
2          Executive        17090.0  
3          Executive        17909.0  
4          Executive        18468.0  
...              ...            ...  
4883         Manager        26576.0  
4884       Exe

## Data Preprocessing

### Merubah Tipe Data

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4888 entries, 0 to 4887
Data columns (total 20 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CustomerID                4888 non-null   int64  
 1   ProdTaken                 4888 non-null   int64  
 2   Age                       4662 non-null   float64
 3   TypeofContact             4863 non-null   object 
 4   CityTier                  4888 non-null   int64  
 5   DurationOfPitch           4637 non-null   float64
 6   Occupation                4888 non-null   object 
 7   Gender                    4888 non-null   object 
 8   NumberOfPersonVisiting    4888 non-null   int64  
 9   NumberOfFollowups         4843 non-null   float64
 10  ProductPitched            4888 non-null   object 
 11  PreferredPropertyStar     4862 non-null   float64
 12  MaritalStatus             4888 non-null   object 
 13  NumberOfTrips             4748 non-null   float64
 14  Passport

In [4]:
df['ProdTaken'] = df['ProdTaken'].astype('str')
df['CityTier'] = df['CityTier'].astype('str')
df['PreferredPropertyStar'] = df['PreferredPropertyStar'].astype('str')
df['Passport'] = df['Passport'].astype('str')
df['PitchSatisfactionScore'] = df['PitchSatisfactionScore'].astype('str')
df['OwnCar'] = df['OwnCar'].astype('str')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4888 entries, 0 to 4887
Data columns (total 20 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CustomerID                4888 non-null   int64  
 1   ProdTaken                 4888 non-null   object 
 2   Age                       4662 non-null   float64
 3   TypeofContact             4863 non-null   object 
 4   CityTier                  4888 non-null   object 
 5   DurationOfPitch           4637 non-null   float64
 6   Occupation                4888 non-null   object 
 7   Gender                    4888 non-null   object 
 8   NumberOfPersonVisiting    4888 non-null   int64  
 9   NumberOfFollowups         4843 non-null   float64
 10  ProductPitched            4888 non-null   object 
 11  PreferredPropertyStar     4888 non-null   object 
 12  MaritalStatus             4888 non-null   object 
 13  NumberOfTrips             4748 non-null   float64
 14  Passport

In [6]:
cats = ['TypeofContact','Occupation','Gender','ProductPitched','MaritalStatus','Designation', 'ProdTaken', 'CityTier', 'PreferredPropertyStar', 'Passport','PitchSatisfactionScore','OwnCar']
nums = ['CustomerID','Age','DurationOfPitch','NumberOfPersonVisiting','NumberOfFollowups','NumberOfTrips','NumberOfChildrenVisiting','MonthlyIncome']

In [7]:
df[nums].describe()

CustomerID          Age  DurationOfPitch  NumberOfPersonVisiting  \
count    4888.000000  4662.000000      4637.000000             4888.000000   
mean   202443.500000    37.622265        15.490835                2.905074   
std      1411.188388     9.316387         8.519643                0.724891   
min    200000.000000    18.000000         5.000000                1.000000   
25%    201221.750000    31.000000         9.000000                2.000000   
50%    202443.500000    36.000000        13.000000                3.000000   
75%    203665.250000    44.000000        20.000000                3.000000   
max    204887.000000    61.000000       127.000000                5.000000   

       NumberOfFollowups  NumberOfTrips  NumberOfChildrenVisiting  \
count        4843.000000    4748.000000               4822.000000   
mean            3.708445       3.236521                  1.187267   
std             1.002509       1.849019                  0.857861   
min             1.000000       1.000000                  0.000000   
25%             3.000000       2.000000                  1.000000   
50%             4.000000       3.000000                  1.000000   
75%             4.000000       4.000000                  2.000000   
max             6.000000      22.000000                  3.000000   

       MonthlyIncome  
count    4655.000000  
mean    23619.853491  
std      5380.698361  
min      1000.000000  
25%     20346.000000  
50%     22347.000000  
75%     25571.000000  
max     98678.000000

In [8]:
df[cats].describe()

TypeofContact Occupation Gender ProductPitched MaritalStatus  \
count           4863       4888   4888           4888          4888   
unique             2          4      3              5             4   
top     Self Enquiry   Salaried   Male          Basic       Married   
freq            3444       2368   2916           1842          2340   

       Designation ProdTaken CityTier PreferredPropertyStar Passport  \
count         4888      4888     4888                  4888     4888   
unique           5         2        3                     4        2   
top      Executive         0        1                   3.0        0   
freq          1842      3968     3190                  2993     3466   

       PitchSatisfactionScore OwnCar  
count                    4888   4888  
unique                      5      2  
top                         3      1  
freq                     1478   3032

### Memperbaiki Fitur Gender

In [9]:
df.replace(to_replace=r'Fe Male', value='Female', regex=True, inplace=True )

In [10]:
df[cats].describe()

TypeofContact Occupation Gender ProductPitched MaritalStatus  \
count           4863       4888   4888           4888          4888   
unique             2          4      2              5             4   
top     Self Enquiry   Salaried   Male          Basic       Married   
freq            3444       2368   2916           1842          2340   

       Designation ProdTaken CityTier PreferredPropertyStar Passport  \
count         4888      4888     4888                  4888     4888   
unique           5         2        3                     4        2   
top      Executive         0        1                   3.0        0   
freq          1842      3968     3190                  2993     3466   

       PitchSatisfactionScore OwnCar  
count                    4888   4888  
unique                      5      2  
top                         3      1  
freq                     1478   3032

### Missing Value

In [11]:
df.isna().sum()

CustomerID                    0
ProdTaken                     0
Age                         226
TypeofContact                25
CityTier                      0
DurationOfPitch             251
Occupation                    0
Gender                        0
NumberOfPersonVisiting        0
NumberOfFollowups            45
ProductPitched                0
PreferredPropertyStar         0
MaritalStatus                 0
NumberOfTrips               140
Passport                      0
PitchSatisfactionScore        0
OwnCar                        0
NumberOfChildrenVisiting     66
Designation                   0
MonthlyIncome               233
dtype: int64

In [12]:
df.head()

CustomerID ProdTaken   Age    TypeofContact CityTier  DurationOfPitch  \
0      200000         1  41.0     Self Enquiry        3              6.0   
1      200001         0  49.0  Company Invited        1             14.0   
2      200002         1  37.0     Self Enquiry        1              8.0   
3      200003         0  33.0  Company Invited        1              9.0   
4      200004         0   NaN     Self Enquiry        1              8.0   

       Occupation  Gender  NumberOfPersonVisiting  NumberOfFollowups  \
0        Salaried  Female                       3                3.0   
1        Salaried    Male                       3                4.0   
2     Free Lancer    Male                       3                4.0   
3        Salaried  Female                       2                3.0   
4  Small Business    Male                       2                3.0   

  ProductPitched PreferredPropertyStar MaritalStatus  NumberOfTrips Passport  \
0         Deluxe                   3.0        Single            1.0        1   
1         Deluxe                   4.0      Divorced            2.0        0   
2          Basic                   3.0        Single            7.0        1   
3          Basic                   3.0      Divorced            2.0        1   
4          Basic                   4.0      Divorced            1.0        0   

  PitchSatisfactionScore OwnCar  NumberOfChildrenVisiting Designation  \
0                      2      1                       0.0     Manager   
1                      3      1                       2.0     Manager   
2                      3      0                       0.0   Executive   
3                      5      1                       1.0   Executive   
4                      5      1                       0.0   Executive   

   MonthlyIncome  
0        20993.0  
1        20130.0  
2        17090.0  
3        17909.0  
4        18468.0

In [13]:
df.dropna(inplace=True, subset=['TypeofContact','NumberOfChildrenVisiting'])
df.isna().sum()

CustomerID                    0
ProdTaken                     0
Age                         226
TypeofContact                 0
CityTier                      0
DurationOfPitch             226
Occupation                    0
Gender                        0
NumberOfPersonVisiting        0
NumberOfFollowups            45
ProductPitched                0
PreferredPropertyStar         0
MaritalStatus                 0
NumberOfTrips               140
Passport                      0
PitchSatisfactionScore        0
OwnCar                        0
NumberOfChildrenVisiting      0
Designation                   0
MonthlyIncome               208
dtype: int64

In [14]:
df['DurationOfPitch'].fillna(df['DurationOfPitch'].median(), inplace=True)
df['NumberOfFollowups'].fillna(df['NumberOfFollowups'].median(), inplace=True)
df['MonthlyIncome'].fillna(df['MonthlyIncome'].median(), inplace=True)

In [15]:
df['Age'].fillna(df['Age'].mode()[0], inplace=True)
df['NumberOfTrips'].fillna(df['NumberOfTrips'].mode()[0], inplace=True)

In [16]:
df.isna().sum()

CustomerID                  0
ProdTaken                   0
Age                         0
TypeofContact               0
CityTier                    0
DurationOfPitch             0
Occupation                  0
Gender                      0
NumberOfPersonVisiting      0
NumberOfFollowups           0
ProductPitched              0
PreferredPropertyStar       0
MaritalStatus               0
NumberOfTrips               0
Passport                    0
PitchSatisfactionScore      0
OwnCar                      0
NumberOfChildrenVisiting    0
Designation                 0
MonthlyIncome               0
dtype: int64

### Duplicated Data

In [17]:
df.duplicated().sum()

0

### Outliers

Menggunakan IQR

Menggunakan Z-score

In [18]:
from scipy import stats
print(f'Jumlah baris sebelum memfilter outlier: {len(df)}')

filtered_entries = np.array([True] * len(df))

for col in ['DurationOfPitch','NumberOfPersonVisiting','NumberOfFollowups','NumberOfTrips','MonthlyIncome']:
    zscore = abs(stats.zscore(df[col])) # hitung absolute z-scorenya
    filtered_entries = (zscore < 3) & filtered_entries # keep yang kurang dari 3 absolute z-scorenya
    
df = df[filtered_entries] # filter, cuma ambil yang z-scorenya dibawah 3

print(f'Jumlah baris setelah memfilter outlier: {len(df)}')

Jumlah baris sebelum memfilter outlier: 4797
Jumlah baris setelah memfilter outlier: 4787


Digunakan Z-score karena terlalu banyak data yang harus didrop jika menggunakan kriteria outlier IQR 

### Normalization/Standardization

In [19]:
df.describe()

CustomerID          Age  DurationOfPitch  NumberOfPersonVisiting  \
count    4787.000000  4787.000000      4787.000000             4787.000000   
mean   202445.630457    37.396490        15.337790                2.906204   
std      1412.764584     9.066171         8.017699                0.725755   
min    200000.000000    18.000000         5.000000                1.000000   
25%    201225.500000    31.000000         9.000000                2.000000   
50%    202446.000000    36.000000        13.000000                3.000000   
75%    203673.500000    43.000000        19.000000                3.000000   
max    204887.000000    61.000000        36.000000                5.000000   

       NumberOfFollowups  NumberOfTrips  NumberOfChildrenVisiting  \
count        4787.000000    4787.000000               4787.000000   
mean            3.710257       3.176729                  1.190934   
std             0.998026       1.765771                  0.858025   
min             1.000000       1.000000                  0.000000   
25%             3.000000       2.000000                  1.000000   
50%             4.000000       3.000000                  1.000000   
75%             4.000000       4.000000                  2.000000   
max             6.000000       8.000000                  3.000000   

       MonthlyIncome  
count    4787.000000  
mean    23382.838730  
std      4892.611455  
min     16009.000000  
25%     20447.500000  
50%     22222.000000  
75%     25274.000000  
max     38677.000000

- min max scaler tidak mengubah distribusi data
- boleh setelah min max lanjut standardscaler
- transformasi log

In [20]:
from sklearn.preprocessing import MinMaxScaler, StandardScaler
scaler = StandardScaler()

df['NumberOfTrips_std'] = StandardScaler().fit_transform(df['NumberOfTrips'].values.reshape(len(df), 1))
df['MonthlyIncome_std'] = StandardScaler().fit_transform(df['MonthlyIncome'].values.reshape(len(df), 1))
df['DurationOfPitch_std'] = scaler.fit_transform(df['DurationOfPitch'].values.reshape(len(df), 1))

<ipython-input-20-5d8f9e92fe67>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['NumberOfTrips_std'] = StandardScaler().fit_transform(df['NumberOfTrips'].values.reshape(len(df), 1))
<ipython-input-20-5d8f9e92fe67>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['MonthlyIncome_std'] = StandardScaler().fit_transform(df['MonthlyIncome'].values.reshape(len(df), 1))
<ipython-input-20-5d8f9e92fe67>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .lo

In [21]:
df.describe()

CustomerID          Age  DurationOfPitch  NumberOfPersonVisiting  \
count    4787.000000  4787.000000      4787.000000             4787.000000   
mean   202445.630457    37.396490        15.337790                2.906204   
std      1412.764584     9.066171         8.017699                0.725755   
min    200000.000000    18.000000         5.000000                1.000000   
25%    201225.500000    31.000000         9.000000                2.000000   
50%    202446.000000    36.000000        13.000000                3.000000   
75%    203673.500000    43.000000        19.000000                3.000000   
max    204887.000000    61.000000        36.000000                5.000000   

       NumberOfFollowups  NumberOfTrips  NumberOfChildrenVisiting  \
count        4787.000000    4787.000000               4787.000000   
mean            3.710257       3.176729                  1.190934   
std             0.998026       1.765771                  0.858025   
min             1.000000       1.000000                  0.000000   
25%             3.000000       2.000000                  1.000000   
50%             4.000000       3.000000                  1.000000   
75%             4.000000       4.000000                  2.000000   
max             6.000000       8.000000                  3.000000   

       MonthlyIncome  NumberOfTrips_std  MonthlyIncome_std  \
count    4787.000000       4.787000e+03       4.787000e+03   
mean    23382.838730      -1.956562e-15      -3.988581e-16   
std      4892.611455       1.000104e+00       1.000104e+00   
min     16009.000000      -1.232865e+00      -1.507295e+00   
25%     20447.500000      -6.664805e-01      -6.000160e-01   
50%     22222.000000      -1.000963e-01      -2.372884e-01   
75%     25274.000000       4.662879e-01       3.865745e-01   
max     38677.000000       2.731825e+00       3.126298e+00   

       DurationOfPitch_std  
count         4.787000e+03  
mean         -1.974954e-16  
std           1.000104e+00  
min          -1.289506e+00  
25%          -7.905575e-01  
50%          -2.916091e-01  
75%           4.568135e-01  
max           2.577344e+00

In [22]:
df.drop(columns=['DurationOfPitch','NumberOfTrips','MonthlyIncome'], inplace=True)
df.head()

C:\Users\ASUS\anaconda3\lib\site-packages\pandas\core\frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


CustomerID ProdTaken   Age    TypeofContact CityTier      Occupation  \
0      200000         1  41.0     Self Enquiry        3        Salaried   
1      200001         0  49.0  Company Invited        1        Salaried   
2      200002         1  37.0     Self Enquiry        1     Free Lancer   
3      200003         0  33.0  Company Invited        1        Salaried   
4      200004         0  35.0     Self Enquiry        1  Small Business   

   Gender  NumberOfPersonVisiting  NumberOfFollowups ProductPitched  \
0  Female                       3                3.0         Deluxe   
1    Male                       3                4.0         Deluxe   
2    Male                       3                4.0          Basic   
3  Female                       2                3.0          Basic   
4    Male                       2                3.0          Basic   

  PreferredPropertyStar MaritalStatus Passport PitchSatisfactionScore OwnCar  \
0                   3.0        Single        1                      2      1   
1                   4.0      Divorced        0                      3      1   
2                   3.0        Single        1                      3      0   
3                   3.0      Divorced        1                      5      1   
4                   4.0      Divorced        0                      5      1   

   NumberOfChildrenVisiting Designation  NumberOfTrips_std  MonthlyIncome_std  \
0                       0.0     Manager          -1.232865          -0.488510   
1                       2.0     Manager          -0.666480          -0.664917   
2                       0.0   Executive           2.165440          -1.286327   
3                       1.0   Executive          -0.666480          -1.118914   
4                       0.0   Executive          -1.232865          -1.004648   

   DurationOfPitch_std  
0            -1.164769  
1            -0.166872  
2            -0.915295  
3            -0.790558  
4            -0.915295

In [23]:
df.describe()

CustomerID          Age  NumberOfPersonVisiting  NumberOfFollowups  \
count    4787.000000  4787.000000             4787.000000        4787.000000   
mean   202445.630457    37.396490                2.906204           3.710257   
std      1412.764584     9.066171                0.725755           0.998026   
min    200000.000000    18.000000                1.000000           1.000000   
25%    201225.500000    31.000000                2.000000           3.000000   
50%    202446.000000    36.000000                3.000000           4.000000   
75%    203673.500000    43.000000                3.000000           4.000000   
max    204887.000000    61.000000                5.000000           6.000000   

       NumberOfChildrenVisiting  NumberOfTrips_std  MonthlyIncome_std  \
count               4787.000000       4.787000e+03       4.787000e+03   
mean                   1.190934      -1.956562e-15      -3.988581e-16   
std                    0.858025       1.000104e+00       1.000104e+00   
min                    0.000000      -1.232865e+00      -1.507295e+00   
25%                    1.000000      -6.664805e-01      -6.000160e-01   
50%                    1.000000      -1.000963e-01      -2.372884e-01   
75%                    2.000000       4.662879e-01       3.865745e-01   
max                    3.000000       2.731825e+00       3.126298e+00   

       DurationOfPitch_std  
count         4.787000e+03  
mean         -1.974954e-16  
std           1.000104e+00  
min          -1.289506e+00  
25%          -7.905575e-01  
50%          -2.916091e-01  
75%           4.568135e-01  
max           2.577344e+00

### Feature encoding

In [24]:
df[cats].describe()

TypeofContact Occupation Gender ProductPitched MaritalStatus  \
count           4787       4787   4787           4787          4787   
unique             2          4      2              5             4   
top     Self Enquiry   Salaried   Male          Basic       Married   
freq            3383       2317   2869           1825          2294   

       Designation ProdTaken CityTier PreferredPropertyStar Passport  \
count         4787      4787     4787                  4787     4787   
unique           5         2        3                     4        2   
top      Executive         0        1                   3.0        0   
freq          1825      3879     3124                  2937     3389   

       PitchSatisfactionScore OwnCar  
count                    4787   4787  
unique                      5      2  
top                         3      1  
freq                     1445   2961

#### Label Encoding

In [25]:
df['TypeofContact_label'] = df['TypeofContact'].astype('category').cat.codes
df['Passport_label'] = df['Passport'].astype('category').cat.codes
df['OwnCar_label'] = df['OwnCar'].astype('category').cat.codes
df['Gender_label'] = df['Gender'].astype('category').cat.codes
df['PreferredPropertyStar_label'] = df['PreferredPropertyStar'].astype('category').cat.codes
df['PitchSatisfactionScore_label'] = df['PitchSatisfactionScore'].astype('category').cat.codes
df['CityTier_label'] = df['CityTier'].astype('category').cat.codes

<ipython-input-25-86ed11dab95e>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['TypeofContact_label'] = df['TypeofContact'].astype('category').cat.codes
<ipython-input-25-86ed11dab95e>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Passport_label'] = df['Passport'].astype('category').cat.codes
<ipython-input-25-86ed11dab95e>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the

In [26]:
df.head()

CustomerID ProdTaken   Age    TypeofContact CityTier      Occupation  \
0      200000         1  41.0     Self Enquiry        3        Salaried   
1      200001         0  49.0  Company Invited        1        Salaried   
2      200002         1  37.0     Self Enquiry        1     Free Lancer   
3      200003         0  33.0  Company Invited        1        Salaried   
4      200004         0  35.0     Self Enquiry        1  Small Business   

   Gender  NumberOfPersonVisiting  NumberOfFollowups ProductPitched  ...  \
0  Female                       3                3.0         Deluxe  ...   
1    Male                       3                4.0         Deluxe  ...   
2    Male                       3                4.0          Basic  ...   
3  Female                       2                3.0          Basic  ...   
4    Male                       2                3.0          Basic  ...   

  NumberOfTrips_std MonthlyIncome_std DurationOfPitch_std TypeofContact_label  \
0         -1.232865         -0.488510           -1.164769                   1   
1         -0.666480         -0.664917           -0.166872                   0   
2          2.165440         -1.286327           -0.915295                   1   
3         -0.666480         -1.118914           -0.790558                   0   
4         -1.232865         -1.004648           -0.915295                   1   

  Passport_label  OwnCar_label Gender_label  PreferredPropertyStar_label  \
0              1             1            0                            0   
1              0             1            1                            1   
2              1             0            1                            0   
3              1             1            0                            0   
4              0             1            1                            1   

   PitchSatisfactionScore_label  CityTier_label  
0                             1               2  
1                             2               0  
2                             2               0  
3                             4               0  
4                             4               0  

[5 rows x 27 columns]

#### One Hot Encoding

In [27]:
col = ['Occupation','ProductPitched','MaritalStatus','Designation']
for cat in col:
    onehots = pd.get_dummies(df[cat], prefix=cat)
    df = df.join(onehots)

In [28]:
df.head()

CustomerID ProdTaken   Age    TypeofContact CityTier      Occupation  \
0      200000         1  41.0     Self Enquiry        3        Salaried   
1      200001         0  49.0  Company Invited        1        Salaried   
2      200002         1  37.0     Self Enquiry        1     Free Lancer   
3      200003         0  33.0  Company Invited        1        Salaried   
4      200004         0  35.0     Self Enquiry        1  Small Business   

   Gender  NumberOfPersonVisiting  NumberOfFollowups ProductPitched  ...  \
0  Female                       3                3.0         Deluxe  ...   
1    Male                       3                4.0         Deluxe  ...   
2    Male                       3                4.0          Basic  ...   
3  Female                       2                3.0          Basic  ...   
4    Male                       2                3.0          Basic  ...   

  ProductPitched_Super Deluxe MaritalStatus_Divorced MaritalStatus_Married  \
0                           0                      0                     0   
1                           0                      1                     0   
2                           0                      0                     0   
3                           0                      1                     0   
4                           0                      1                     0   

  MaritalStatus_Single MaritalStatus_Unmarried  Designation_AVP  \
0                    1                       0                0   
1                    0                       0                0   
2                    1                       0                0   
3                    0                       0                0   
4                    0                       0                0   

  Designation_Executive  Designation_Manager  Designation_Senior Manager  \
0                     0                    1                           0   
1                     0                    1                           0   
2                     1                    0                           0   
3                     1                    0                           0   
4                     1                    0                           0   

   Designation_VP  
0               0  
1               0  
2               0  
3               0  
4               0  

[5 rows x 45 columns]

In [29]:
df.drop(columns=['Occupation','Gender','ProductPitched','MaritalStatus','Designation', 'CityTier', 
                 'PreferredPropertyStar', 'PitchSatisfactionScore', 'TypeofContact', 'Passport', 'OwnCar'], inplace=True)
df

CustomerID ProdTaken   Age  NumberOfPersonVisiting  NumberOfFollowups  \
0         200000         1  41.0                       3                3.0   
1         200001         0  49.0                       3                4.0   
2         200002         1  37.0                       3                4.0   
3         200003         0  33.0                       2                3.0   
4         200004         0  35.0                       2                3.0   
...          ...       ...   ...                     ...                ...   
4883      204883         1  49.0                       3                5.0   
4884      204884         1  28.0                       4                5.0   
4885      204885         1  52.0                       4                4.0   
4886      204886         1  19.0                       3                4.0   
4887      204887         1  36.0                       4                4.0   

      NumberOfChildrenVisiting  NumberOfTrips_std  MonthlyIncome_std  \
0                          0.0          -1.232865          -0.488510   
1                          2.0          -0.666480          -0.664917   
2                          0.0           2.165440          -1.286327   
3                          1.0          -0.666480          -1.118914   
4                          0.0          -1.232865          -1.004648   
...                        ...                ...                ...   
4883                       1.0          -0.666480           0.652718   
4884                       2.0          -0.100096          -0.443744   
4885                       3.0           2.165440           1.724650   
4886                       2.0          -0.100096          -0.632415   
4887                       2.0          -0.100096           0.134536   

      DurationOfPitch_std  TypeofContact_label  ...  \
0               -1.164769                    1  ...   
1               -0.166872                    0  ...   
2               -0.915295                    1  ...   
3               -0.790558                    0  ...   
4               -0.915295                    1  ...   
...                   ...                  ...  ...   
4883            -0.790558                    1  ...   
4884             1.953659                    0  ...   
4885             0.207339                    1  ...   
4886             0.082602                    1  ...   
4887            -0.166872                    1  ...   

      ProductPitched_Super Deluxe  MaritalStatus_Divorced  \
0                               0                       0   
1                               0                       1   
2                               0                       0   
3                               0                       1   
4                               0                       1   
...                           ...                     ...   
4883                            0                       0   
4884                            0                       0   
4885                            0                       0   
4886                            0                       0   
4887                            0                       0   

      MaritalStatus_Married  MaritalStatus_Single  MaritalStatus_Unmarried  \
0                         0                     1                        0   
1                         0                     0                        0   
2                         0                     1                        0   
3                         0                     0                        0   
4                         0                     0                        0   
...                     ...                   ...                      ...   
4883                      0                     0                        1   
4884                      0                     1                        0   
4885                      1                     0                        0   
4886              

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4787 entries, 0 to 4887
Data columns (total 34 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   CustomerID                    4787 non-null   int64  
 1   ProdTaken                     4787 non-null   object 
 2   Age                           4787 non-null   float64
 3   NumberOfPersonVisiting        4787 non-null   int64  
 4   NumberOfFollowups             4787 non-null   float64
 5   NumberOfChildrenVisiting      4787 non-null   float64
 6   NumberOfTrips_std             4787 non-null   float64
 7   MonthlyIncome_std             4787 non-null   float64
 8   DurationOfPitch_std           4787 non-null   float64
 9   TypeofContact_label           4787 non-null   int8   
 10  Passport_label                4787 non-null   int8   
 11  OwnCar_label                  4787 non-null   int8   
 12  Gender_label                  4787 non-null   int8   
 13  Pre

## Modeling

### Membagi Data Test dan Data Training

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4787 entries, 0 to 4887
Data columns (total 34 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   CustomerID                    4787 non-null   int64  
 1   ProdTaken                     4787 non-null   object 
 2   Age                           4787 non-null   float64
 3   NumberOfPersonVisiting        4787 non-null   int64  
 4   NumberOfFollowups             4787 non-null   float64
 5   NumberOfChildrenVisiting      4787 non-null   float64
 6   NumberOfTrips_std             4787 non-null   float64
 7   MonthlyIncome_std             4787 non-null   float64
 8   DurationOfPitch_std           4787 non-null   float64
 9   TypeofContact_label           4787 non-null   int8   
 10  Passport_label                4787 non-null   int8   
 11  OwnCar_label                  4787 non-null   int8   
 12  Gender_label                  4787 non-null   int8   
 13  Pre

In [32]:
kol = ['Occupation_Free Lancer', 'Occupation_Large Business',
       'Occupation_Salaried', 'Occupation_Small Business',
       'ProductPitched_Basic', 'ProductPitched_Deluxe', 'ProductPitched_King',
       'ProductPitched_Standard', 'ProductPitched_Super Deluxe',
       'MaritalStatus_Divorced', 'MaritalStatus_Married',
       'MaritalStatus_Single', 'MaritalStatus_Unmarried', 'Designation_AVP',
       'Designation_Executive', 'Designation_Manager',
       'Designation_Senior Manager', 'Designation_VP']
for i in kol:
    df[i] = df[i].astype(int)

df['ProdTaken'] = df['ProdTaken'].astype(int)

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4787 entries, 0 to 4887
Data columns (total 34 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   CustomerID                    4787 non-null   int64  
 1   ProdTaken                     4787 non-null   int32  
 2   Age                           4787 non-null   float64
 3   NumberOfPersonVisiting        4787 non-null   int64  
 4   NumberOfFollowups             4787 non-null   float64
 5   NumberOfChildrenVisiting      4787 non-null   float64
 6   NumberOfTrips_std             4787 non-null   float64
 7   MonthlyIncome_std             4787 non-null   float64
 8   DurationOfPitch_std           4787 non-null   float64
 9   TypeofContact_label           4787 non-null   int8   
 10  Passport_label                4787 non-null   int8   
 11  OwnCar_label                  4787 non-null   int8   
 12  Gender_label                  4787 non-null   int8   
 13  Pre

In [33]:
from sklearn.model_selection import train_test_split

X = df[list(df.columns[2:])]
Y = df[['ProdTaken']]
X_train, X_test, y_train, y_test = train_test_split(X,Y,test_size = 0.2,random_state = 42)

In [34]:
df.shape

(4787, 34)

In [35]:
X_train.shape

(3829, 32)

In [36]:
X_test.shape

(958, 32)

## Data Split Conclusion

- Data training : 3829 rows x 32 columns
- Data testing : 958 rows x 32 columns

### Modeling XGBoost

In [37]:
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

def eval_classification(model, pred, xtrain, ytrain, xtest, ytest):
    print("Accuracy (Test Set): %.2f" % accuracy_score(ytest, pred))
    print("Precision (Test Set): %.2f" % precision_score(ytest, pred))
    print("Recall (Test Set): %.2f" % recall_score(ytest, pred))
    print("F1-Score (Test Set): %.2f" % f1_score(ytest, pred))
    
xg = XGBClassifier(random_state=42)
xg.fit(X_train, y_train)

y_pred_test = xg.predict(X_test)
eval_classification(xg, y_pred_test, X_train, y_train, X_test, y_test)

C:\Users\ASUS\anaconda3\lib\site-packages\xgboost\sklearn.py:1224: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)
C:\Users\ASUS\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


[15:17:01] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.5.1/src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
Accuracy (Test Set): 0.93
Precision (Test Set): 0.95
Recall (Test Set): 0.70
F1-Score (Test Set): 0.81


In [38]:
from sklearn.metrics import confusion_matrix

cf = confusion_matrix(y_test, y_pred_test)
cf

array([[752,   7],
       [ 59, 140]], dtype=int64)

In [39]:
from sklearn.metrics import precision_score
xg.fit(X_train, y_train)
y_pred_train = xg.predict(X_train)
y_pred = xg.predict(X_test)
print("Precision (Train Set): " +str(precision_score(y_train, y_pred_train)))
print("Precision (Test Set):" +str(precision_score(y_test, y_pred_test)))

[15:17:02] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.5.1/src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


C:\Users\ASUS\anaconda3\lib\site-packages\xgboost\sklearn.py:1224: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)
C:\Users\ASUS\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


Precision (Train Set): 1.0
Precision (Test Set):0.9523809523809523


## Data for Business Metrics Calculation

Data testing : 958 rows

In [40]:
df

CustomerID  ProdTaken   Age  NumberOfPersonVisiting  NumberOfFollowups  \
0         200000          1  41.0                       3                3.0   
1         200001          0  49.0                       3                4.0   
2         200002          1  37.0                       3                4.0   
3         200003          0  33.0                       2                3.0   
4         200004          0  35.0                       2                3.0   
...          ...        ...   ...                     ...                ...   
4883      204883          1  49.0                       3                5.0   
4884      204884          1  28.0                       4                5.0   
4885      204885          1  52.0                       4                4.0   
4886      204886          1  19.0                       3                4.0   
4887      204887          1  36.0                       4                4.0   

      NumberOfChildrenVisiting  NumberOfTrips_std  MonthlyIncome_std  \
0                          0.0          -1.232865          -0.488510   
1                          2.0          -0.666480          -0.664917   
2                          0.0           2.165440          -1.286327   
3                          1.0          -0.666480          -1.118914   
4                          0.0          -1.232865          -1.004648   
...                        ...                ...                ...   
4883                       1.0          -0.666480           0.652718   
4884                       2.0          -0.100096          -0.443744   
4885                       3.0           2.165440           1.724650   
4886                       2.0          -0.100096          -0.632415   
4887                       2.0          -0.100096           0.134536   

      DurationOfPitch_std  TypeofContact_label  ...  \
0               -1.164769                    1  ...   
1               -0.166872                    0  ...   
2               -0.915295                    1  ...   
3               -0.790558                    0  ...   
4               -0.915295                    1  ...   
...                   ...                  ...  ...   
4883            -0.790558                    1  ...   
4884             1.953659                    0  ...   
4885             0.207339                    1  ...   
4886             0.082602                    1  ...   
4887            -0.166872                    1  ...   

      ProductPitched_Super Deluxe  MaritalStatus_Divorced  \
0                               0                       0   
1                               0                       1   
2                               0                       0   
3                               0                       1   
4                               0                       1   
...                           ...                     ...   
4883                            0                       0   
4884                            0                       0   
4885                            0                       0   
4886                            0                       0   
4887                            0                       0   

      MaritalStatus_Married  MaritalStatus_Single  MaritalStatus_Unmarried  \
0                         0                     1                        0   
1                         0                     0                        0   
2                         0                     1                        0   
3                         0                     0                        0   
4                         0                     0                        0   
...                     ...                   ...                      ...   
4883                      0                     0                        1   
4884                      0                     1                        0   
4885                      1                     0                        0   
4886  

In [41]:
df['DurationOfPitch'] = scaler.inverse_transform(df['DurationOfPitch_std'])
df

CustomerID  ProdTaken   Age  NumberOfPersonVisiting  NumberOfFollowups  \
0         200000          1  41.0                       3                3.0   
1         200001          0  49.0                       3                4.0   
2         200002          1  37.0                       3                4.0   
3         200003          0  33.0                       2                3.0   
4         200004          0  35.0                       2                3.0   
...          ...        ...   ...                     ...                ...   
4883      204883          1  49.0                       3                5.0   
4884      204884          1  28.0                       4                5.0   
4885      204885          1  52.0                       4                4.0   
4886      204886          1  19.0                       3                4.0   
4887      204887          1  36.0                       4                4.0   

      NumberOfChildrenVisiting  NumberOfTrips_std  MonthlyIncome_std  \
0                          0.0          -1.232865          -0.488510   
1                          2.0          -0.666480          -0.664917   
2                          0.0           2.165440          -1.286327   
3                          1.0          -0.666480          -1.118914   
4                          0.0          -1.232865          -1.004648   
...                        ...                ...                ...   
4883                       1.0          -0.666480           0.652718   
4884                       2.0          -0.100096          -0.443744   
4885                       3.0           2.165440           1.724650   
4886                       2.0          -0.100096          -0.632415   
4887                       2.0          -0.100096           0.134536   

      DurationOfPitch_std  TypeofContact_label  ...  MaritalStatus_Divorced  \
0               -1.164769                    1  ...                       0   
1               -0.166872                    0  ...                       1   
2               -0.915295                    1  ...                       0   
3               -0.790558                    0  ...                       1   
4               -0.915295                    1  ...                       1   
...                   ...                  ...  ...                     ...   
4883            -0.790558                    1  ...                       0   
4884             1.953659                    0  ...                       0   
4885             0.207339                    1  ...                       0   
4886             0.082602                    1  ...                       0   
4887            -0.166872                    1  ...                       0   

      MaritalStatus_Married  MaritalStatus_Single  MaritalStatus_Unmarried  \
0                         0                     1                        0   
1                         0                     0                        0   
2                         0                     1                        0   
3                         0                     0                        0   
4                         0                     0                        0   
...                     ...                   ...                      ...   
4883                      0                     0                        1   
4884                      0                     1                        0   
4885                      1                     0                        0   
4886                      0                     1                        0   
4887                      0                     0                        1   

      Designation_AVP  Designation_Executive  Designation_Manager  \
0                   0                      0                    1   
1                   0                      0                    1   
2                   0                      1                    0   
3                 

In [42]:
X_test

Age  NumberOfPersonVisiting  NumberOfFollowups  \
567   34.0                       2                3.0   
913   45.0                       3                5.0   
1722  29.0                       2                3.0   
718   35.0                       2                3.0   
1911  42.0                       3                3.0   
...    ...                     ...                ...   
1805  40.0                       2                5.0   
1897  60.0                       3                4.0   
4373  42.0                       4                5.0   
663   21.0                       3                3.0   
2352  36.0                       3                5.0   

      NumberOfChildrenVisiting  NumberOfTrips_std  MonthlyIncome_std  \
567                        0.0          -0.666480          -0.431070   
913                        2.0          -0.100096           2.134290   
1722                       1.0          -0.666480          -1.235224   
718                        0.0          -0.666480          -0.970920   
1911                       1.0           1.032672          -1.285713   
...                        ...                ...                ...   
1805                       0.0          -0.100096          -0.189252   
1897                       0.0           1.032672           0.384939   
4373                       1.0           0.466288           0.430523   
663                        1.0          -0.666480          -1.163067   
2352                       0.0          -0.100096           0.382077   

      DurationOfPitch_std  TypeofContact_label  Passport_label  OwnCar_label  \
567             -0.915295                    1               0             0   
913             -0.915295                    1               1             0   
1722             0.706288                    1               0             1   
718              1.704184                    0               1             1   
1911            -0.541083                    0               0             0   
...                   ...                  ...             ...           ...   
1805            -1.164769                    1               1             1   
1897             2.327870                    0               0             1   
4373            -0.790558                    0               1             0   
663              1.080499                    1               0             1   
2352            -1.040032                    1               1             1   

      ...  ProductPitched_Super Deluxe  MaritalStatus_Divorced  \
567   ...                            0                       0   
913   ...                            0                       0   
1722  ...                            0                       0   
718   ...                            0                       1   
1911  ...                            0                       0   
...   ...                          ...                     ...   
1805  ...                            0                       0   
1897  ...                            0                       0   
4373  ...                            0                       0   
663   ...                            0                       0   
2352  ...                            0                       0   

      MaritalStatus_Married  MaritalStatus_Single  MaritalStatus_Unmarried  \
567                       0                     1                        0   
913                       0                     1                        0   
1722                      0                     1                        0   
718                       0                     0                        0   
1911                      1                     0                        0   
...                     ...                   ...                      ...   
1805                      1                     0                        0   
1897                      1                     0                        0   
4

In [43]:
Data = X_test
Data['ProdTaken'] = y_pred_test
Data

<ipython-input-43-b50d74f3d87c>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Data['ProdTaken'] = y_pred_test


Age  NumberOfPersonVisiting  NumberOfFollowups  \
567   34.0                       2                3.0   
913   45.0                       3                5.0   
1722  29.0                       2                3.0   
718   35.0                       2                3.0   
1911  42.0                       3                3.0   
...    ...                     ...                ...   
1805  40.0                       2                5.0   
1897  60.0                       3                4.0   
4373  42.0                       4                5.0   
663   21.0                       3                3.0   
2352  36.0                       3                5.0   

      NumberOfChildrenVisiting  NumberOfTrips_std  MonthlyIncome_std  \
567                        0.0          -0.666480          -0.431070   
913                        2.0          -0.100096           2.134290   
1722                       1.0          -0.666480          -1.235224   
718                        0.0          -0.666480          -0.970920   
1911                       1.0           1.032672          -1.285713   
...                        ...                ...                ...   
1805                       0.0          -0.100096          -0.189252   
1897                       0.0           1.032672           0.384939   
4373                       1.0           0.466288           0.430523   
663                        1.0          -0.666480          -1.163067   
2352                       0.0          -0.100096           0.382077   

      DurationOfPitch_std  TypeofContact_label  Passport_label  OwnCar_label  \
567             -0.915295                    1               0             0   
913             -0.915295                    1               1             0   
1722             0.706288                    1               0             1   
718              1.704184                    0               1             1   
1911            -0.541083                    0               0             0   
...                   ...                  ...             ...           ...   
1805            -1.164769                    1               1             1   
1897             2.327870                    0               0             1   
4373            -0.790558                    0               1             0   
663              1.080499                    1               0             1   
2352            -1.040032                    1               1             1   

      ...  MaritalStatus_Divorced  MaritalStatus_Married  \
567   ...                       0                      0   
913   ...                       0                      0   
1722  ...                       0                      0   
718   ...                       1                      0   
1911  ...                       0                      1   
...   ...                     ...                    ...   
1805  ...                       0                      1   
1897  ...                       0                      1   
4373  ...                       0                      0   
663   ...                       0                      0   
2352  ...                       0                      1   

      MaritalStatus_Single  MaritalStatus_Unmarried  Designation_AVP  \
567                      1                        0                0   
913                      1                        0                0   
1722                     1                        0                0   
718                      0                        0                0   
1911                     0                        0                0   
...                    ...                      ...              ...   
1805                     0                        0                0   
1897                     0                        0                0   
4373                     0                        1                0   
663                      1                        0    

In [44]:
Data_merge = pd.merge(Data, df['DurationOfPitch'], left_index=True, right_index=True)
Data_merge = Data_merge.reset_index(drop = True)
Data_merge

Age  NumberOfPersonVisiting  NumberOfFollowups  \
0    34.0                       2                3.0   
1    45.0                       3                5.0   
2    29.0                       2                3.0   
3    35.0                       2                3.0   
4    42.0                       3                3.0   
..    ...                     ...                ...   
953  40.0                       2                5.0   
954  60.0                       3                4.0   
955  42.0                       4                5.0   
956  21.0                       3                3.0   
957  36.0                       3                5.0   

     NumberOfChildrenVisiting  NumberOfTrips_std  MonthlyIncome_std  \
0                         0.0          -0.666480          -0.431070   
1                         2.0          -0.100096           2.134290   
2                         1.0          -0.666480          -1.235224   
3                         0.0          -0.666480          -0.970920   
4                         1.0           1.032672          -1.285713   
..                        ...                ...                ...   
953                       0.0          -0.100096          -0.189252   
954                       0.0           1.032672           0.384939   
955                       1.0           0.466288           0.430523   
956                       1.0          -0.666480          -1.163067   
957                       0.0          -0.100096           0.382077   

     DurationOfPitch_std  TypeofContact_label  Passport_label  OwnCar_label  \
0              -0.915295                    1               0             0   
1              -0.915295                    1               1             0   
2               0.706288                    1               0             1   
3               1.704184                    0               1             1   
4              -0.541083                    0               0             0   
..                   ...                  ...             ...           ...   
953            -1.164769                    1               1             1   
954             2.327870                    0               0             1   
955            -0.790558                    0               1             0   
956             1.080499                    1               0             1   
957            -1.040032                    1               1             1   

     ...  MaritalStatus_Married  MaritalStatus_Single  \
0    ...                      0                     1   
1    ...                      0                     1   
2    ...                      0                     1   
3    ...                      0                     0   
4    ...                      1                     0   
..   ...                    ...                   ...   
953  ...                      1                     0   
954  ...                      1                     0   
955  ...                      0                     0   
956  ...                      0                     1   
957  ...                      1                     0   

     MaritalStatus_Unmarried  Designation_AVP  Designation_Executive  \
0                          0                0                      0   
1                          0                0                      0   
2                          0                0                      1   
3                          0                0                      0   
4                          0                0                      1   
..                       ...              ...                    ...   
953                        0                0                      0   
954                        0                0                      0   
955                        1                0                      0   
956                        0                0                      1   
957                        0                0         

## Before Using Model

In [45]:
Basic = Data_merge['ProductPitched_Basic'].sum()
Deluxe = Data_merge['ProductPitched_Deluxe'].sum()
King = Data_merge['ProductPitched_King'].sum()
Standard = Data_merge['ProductPitched_Standard'].sum()
SuperDeluxe = Data_merge['ProductPitched_Super Deluxe'].sum()

#### Expected Revenue

Price Source: https://costaricaexperts.com/package/best/

In [46]:
ExpRev_before = (Basic*1000) + (Standard*2000) + (Deluxe*3000) + (SuperDeluxe*4000) + (King*5000)
ExpRev_before

2201000

#### Telemarketing Cost

Source:
https://www.magellan-solutions.com/blog/cost-of-telemarketing/

In [47]:
Data_merge['TeleCost'] = ((Data_merge['DurationOfPitch'] * Data_merge['NumberOfFollowups'])/60) * 50
TeleCost_before = Data_merge['TeleCost'].sum()
TeleCost_before

46342.5

#### Actual Revenue

In [48]:
ActRev_before = ExpRev_before*0.18
ActRev_before

396180.0

#### Spending Revenue on Telemarketing Cost

'You should spend 2–5% of your sales revenue on marketing'

Source: https://nuphoriq.com/create-a-marketing-budget/

In [49]:
Spending_before = round((TeleCost_before/ActRev_before)*100,2)
Spending_before

11.7

Biaya telemarketing tidak efisien. Spending revenue on telemarketing cost sebesar 11.7%, sedangkan persentase normal sales revenue on marketing sebesar 2-5%.

## After Using Model

Perhitungan tetap menggunakan data test berjumlah 958 rows dengan asumsi jumlah data ini sudah dikalikan dengan recall dari model sehingga didapatkan expected revenues yang sama (dan pada akhirnya cost telemarketing yang dikeluarkan juga sama)

#### Expected Revenue

Price Source: https://costaricaexperts.com/package/best/

In [50]:
ExpRev_after = (Basic*1000) + (Standard*2000) + (Deluxe*3000) + (SuperDeluxe*4000) + (King*5000)
ExpRev_after

2201000

#### Telemarketing Cost

Source:
https://www.magellan-solutions.com/blog/cost-of-telemarketing/

In [51]:
Data_merge['TeleCost'] = ((Data_merge['DurationOfPitch'] * Data_merge['NumberOfFollowups'])/60) * 50
TeleCost_after = Data_merge['TeleCost'].sum()
TeleCost_after

46342.5

#### Actual Revenue

In [52]:
ActRev_after = ExpRev_after*0.95
ActRev_after

2090950.0

#### Spending Revenue on Telemarketing Cost

'You should spend 2–5% of your sales revenue on marketing'

Source: https://nuphoriq.com/create-a-marketing-budget/

In [53]:
Spending_after = round((TeleCost_after/ActRev_after)*100,2)
Spending_after

2.22

### Conclusions

#### Kenaikan Revenue

In [54]:
print(round(((ActRev_after-ActRev_before)/(ActRev_before))*100,1),'persen')

427.8 persen


#### Penurunan Spending Revenue on Telemarketing Cost

In [55]:
print(round((Spending_before-Spending_after),1),'persen')

9.5 persen
